docs/rfcs/0020-rotate-hash-peppers.md

```diff
@@ -0,0 +1,379 @@
+
+ A rough outline of functionality might look like:
+```
+ALTER TABLE BENEFICIARIES
+   ADD COLUMN mbi_hash_standby CHARACTER VARYING(64),
+   ADD COLUMN bene_crnt_hic_num_standby CHARACTER VARYING(64);
+```
+- the _BENEFICIARIES_ table thus contains 2 sets of MBI and HICN data columns
+  - MBI_HASH
+  - BENE_CRNT_HIC_NUM
+  - MBI_HASH_STANDBY
+  - BENE_CRNT_HIC_NUM_STANDBY
+- some sort of script (python, SQL) could then update the _STANDBY columns using the new hash algorithm.
+- when necessary to make the _STANDBY values _active_; a SQL script could be executed to do that; something like:
+```
+UPDATE BENEFICIARIES SET
+  mbi_hash = mbi_hash_standby,
+  bene_crnt_hic_num = bene_crnt_hic_num_standby
+    WHERE bene_id EXISTS
+      AND MBI_HASH_STANDBY IS NOT NULL;
+```
+#### PROS
+- minimal changes to current processing
+  - standby columns need to be added to ORM
+  - some script work (python or SQL)
+- minor impact to PP; they will need to modify their pre-request URL construction to use the new hash algorithm so it would be presumptuous to consider this a _zero downtime_ solution.
+#### CONS
+- same table UPDATE, INDEXES introduces some challenges; may need to re-build index concurrently which can be slower than having table locked for indexing which is feasible, as noted in the prior use case using a separate _BENEFICIARIES_PREHASH_ table.
+- maintaining data over time (i.e., INSERTs or UPDATEs from weekly ETL) needs to be considered; changes to ETL may be non-trivial.
+- may also require some AWS (SSM) work
+- will need script (i.e., flyway, python, SQL) that can populate new table columns; possibly custom maven build artifact.
+### Hot Stand-by : Option 2
+[Hot Stand-by - Option 2]: #hot-stand-by-2
+
+Similar to Option 1, but maintain the _shadow_ values for MBI and HICN in a (new) separate table.
+
+ A rough outline of functionality might look like:
+- SSM parameter store upgraded to support multiple hash algorithm instances:
+  - SSM hash alg key
+  - SSM hash alg salt (seed value)
+  - SSM hash alg iteration count
+  - SSM hash alg status (active, future, disabled)
+- create table for storing concurrent beneficiary hash values; a sample table structure might look like:
+```
+CREATE TABLE IF NOT EXISTS hashed_beneficiaries
+(
+    id                bigint NOT NULL,
+    ssm_param_id      character varying(32) NOT NULL,
+    bene_id           bigint NOT NULL,
+    bene_crnt_hic_num character varying(64) NOT NULL,
+    mbi_hash          character varying(64),
+    efctv_end_dt      date,
+    CONSTRAINT hashed_beneficiaries_pkey PRIMARY KEY (id)
+);
+
+CREATE INDEX IF NOT EXISTS hashed_beneficiaries_ssm_bene_id_idx
+    ON hashed_beneficiaries USING btree (ssm_param_id, bene_id);
+
+CREATE INDEX IF NOT EXISTS hashed_beneficiaries_mbi_hash_idx
+    ON hashed_beneficiaries USING btree (mbi_hash);
+    
+CREATE INDEX IF NOT EXISTS hashed_beneficiaries_crnt_hic_num_idx
+    ON hashed_beneficiaries USING btree (bene_crnt_hic_num);
```

Comment:
Not impossible to wind up with collisions between records for different benes and settings on these hashed values.  Might be vanishingly small probability though.

Reply:
true of pretty much any hashing alg; we already have both those indices; this is just rebuilding them on a temp table. However, using the proposed solution (HTTP POST) transfers hash collision responsibility to the source of truth, CMS who is responsible for MBI(s).